src/coordination/coordinator_instance.cpp

```diff
@@ -214,80 +222,151 @@ auto CoordinatorInstance::ShowInstances() const -> std::vector<InstanceStatus> {
+    if (raft_state_->IsCurrentMain(instance.InstanceName())) return "main";
+    return "replica";
+  };
+
+  auto const stringify_repl_health = [](ReplicationInstanceConnector const &instance) -> std::string {
+    return instance.IsAlive() ? "up" : "down";
+  };
 
-    std::ranges::transform(raft_state_->GetReplicationInstances(), std::back_inserter(instances_status),
-                           process_repl_instance_as_follower);
+  auto process_repl_instance_as_leader =
+      [&stringify_repl_role, &stringify_repl_health](ReplicationInstanceConnector const &instance) -> InstanceStatus {
+    return {.instance_name = instance.InstanceName(),
+            .management_server = instance.ManagementSocketAddress(),  // show non-resolved IP
+            .bolt_server = instance.BoltSocketAddress(),              // show non-resolved IP
+            .cluster_role = stringify_repl_role(instance),
+            .health = stringify_repl_health(instance),
+            .last_succ_resp_ms = instance.LastSuccRespMs().count()};
+  };
+
+  {
+    auto lock = std::shared_lock{coord_instance_lock_};
+    std::ranges::transform(repl_instances_, std::back_inserter(instances_status), process_repl_instance_as_leader);
   }
+
   return instances_status;
 }
 
+auto CoordinatorInstance::ShowInstances() -> std::vector<InstanceStatus> {
+  if (is_leader_ready_) {
+    return ShowInstancesAsLeader();
+  }
+
+  auto const leader_id = raft_state_->GetLeaderId();
+  if (leader_id == raft_state_->GetCoordinatorId()) {
+    spdlog::trace("Coordinator itself not yet leader, returning report as follower.");
+    return ShowInstancesStatusAsFollower();  // We don't want to ask ourselves for instances, as coordinator is
+                                             // not ready still as leader
+  }
+
+  if (leader_id == -1) {
+    spdlog::trace("No leader found, returning report as follower");
+    return ShowInstancesStatusAsFollower();
+  }
+  CoordinatorInstanceConnector *follower{nullptr};
+  // TODO(antoniofilipovic) move this part to separate function
+  {
+    auto connectors = coordinator_connectors_.Lock();
+    if (!connectors->contains(leader_id)) {
+      auto all_coordinators = raft_state_->GetCoordinatorInstances();
+
+      auto coord = std::ranges::find_if(all_coordinators,
+                                        [leader_id](auto const &coord) { return coord.coordinator_id == leader_id; });
+
+      MG_ASSERT(coord != all_coordinators.end());
+      spdlog::trace("Creating new connector for leader with id {}, endpoint {}", leader_id,
+                    coord->management_server.SocketAddress());
+      connectors->emplace(leader_id, ManagementServerConfig{coord->management_server});
+    }
+    follower = &connectors->at(leader_id);
+  }
+
+  MG_ASSERT(follower != nullptr, "Follower not found");
+
+  spdlog::trace("Sending show instances RPC to leader with id {}", leader_id);
+  auto maybe_res = follower->SendShowInstances();
```

Comment:
const

Reply:
Can't be const as I am moving value from it later on, so it would make copy